coastguard/clustersetnetworkpolicy.md

```diff
@@ -0,0 +1,272 @@
+with additional fields that are relevant to the clusterset.
+
+The policies are local to the cluster, meaning they affect only the pods they are targeting on that cluster.
+If users want, they can create the same policy across different clusters manually.
+In the future, we may consider making policies global (or adding such a capability via a new CRD),
+allowing the users to create the policy once, and the system will take care of distributing it.
+
+Each such policy will be translated by the local CoastGuard controller to one (or more) `NetworkPolicy` where:
+
+* Local peers use the selector defined in the `ClusterSetNetworkPolicy` CR.
+* Remote peers (pods) are identified using an `ipBlock` with a `/32` CIDR.
+
+These local policies will be updated by the controller in response to any change in the original `ClusterSetNetworkPolicy` or any remote
+pod or namespace that are relevant to the policy.
+
+#### Proposed CR Structure
+
+```yaml
+apiVersion: submariner.io/v1alpha1
+kind: ClusterSetNetworkPolicy
+metadata:
+  name: <string>
+  namespace: <string>
+spec:
+  podSelector: <LabelSelector>
+  policyTypes: <string array>
+  ingress:
+  - <NetworkPolicyIngressRule>
+  - from:
+    - <NetworkPolicyPeer>
+    - namespaceSelector: <LabelSelector>
+      podSelector: <LabelSelector>
+    ports: <NetworkPolicyPort array>
+  egress:
+  - <NetworkPolicyEgressRule>
+  - to:
+    - <NetworkPolicyPeer>
+    - namespaceSelector: <LabelSelector>
+      podSelector: <LabelSelector>
+    ports: <NetworkPolicyPort array>
+```
+
+### Tracking Remote Pod IPs
+
+#### RemotePod & RemoteNamespace CRDs
+
+A simple approach to tracking remote pods is having a specific CRD representing each remote pod and a CRD representing remote namespaces.
+These CRDs would be namespace-scoped but the namespace it's in will be a system set namespace (e.g. submariner-operator).
+These CRDs will be synchronized via the broker.
+
+Each cluster will track its own pods and namespaces and keep them updated on each change in their lifecycle:
+
+* When a pod/namespace is created, create a corresponding `RemotePod`/`RemoteNamespace`.
+* When a pod/namespace is updated, update the corresponding `RemotePod`/`RemoteNamespace`.
+* When a pod/namespace is deleted, delete the corresponding `RemotePod`/`RemoteNamespace`.
+
+The advantage of this approach is that it's quite simple for each cluster to track and generate these CRs,
+and each consuming cluster will use them for the actual implementation.
+A disadvantage of this approach is that we'll be tracking every pod on every cluster, even if they don't match any selector.
+This can be mitigated somewhat by having the source cluster track only pods/namespaces that match relevant selectors.
+
+##### Proposed RemotePod CR Structure
+
+```yaml
+apiVersion: submariner.io/v1alpha1
+kind: RemotePod
+metadata:
+  labels: <map string -> string>
+  name: <string>
```

Comment:
For every pod in the remote cluster will we have an RemotePod object representing that pod in the local cluster?

I know its an implementation detail, but just to understand, may i know how the objects would be named to avoid any collision?